File: scratch.js
const fs = require('fs');
const file = '/Users/mac/Desktop/pesanest1/src/components/dashboard/ApprovalQueue.tsx';
let content = fs.readFileSync(file, 'utf8');

// Update function signature
content = content.replace(
    /action: 'APPROVE' \| 'REJECT'\) => {/,
    "action: 'APPROVE' | 'REJECT' | 'ADJUST') => {"
);

content = content.replace(
    /decision: action === 'APPROVE' \? 'APPROVED' : 'REJECTED'/,
    "decision: action === 'APPROVE' ? 'APPROVED' : (action === 'REJECT' ? 'REJECTED' : 'ADJUSTMENT')"
);

// We have grid buttons of form: <div className="grid grid-cols-2 gap-2">
const gridRegex = /<div className="grid grid-cols-2 gap-2">\s*<button\s*onClick={\(\) => handleApproval\(\(([^ ]+) as any\)\.approvalId, 'APPROVE'\)}\s*disabled={isProcessing}\s*className="py-2 bg-\[#29258D\] text-white text-xs font-semibold rounded-md hover:bg-\[#29258D\]\/90 transition-all flex items-center justify-center gap-1"\s*>\s*{isProcessing \? <PiSpinner className="animate-spin" \/> : <PiCheckCircle \/>} Approve\s*<\/button>\s*<button\s*onClick={\(\) => handleApproval\(\(\1 as any\)\.approvalId, 'REJECT'\)}\s*disabled={isProcessing}\s*className="py-2 bg-white border border-rose-200 text-rose-600 text-xs font-semibold rounded-md hover:bg-rose-50 transition-all flex items-center justify-center gap-1"\s*>\s*{isProcessing \? <PiSpinner className="animate-spin" \/> : <PiXCircle \/>} Reject\s*<\/button>\s*<\/div>/g;

content = content.replace(gridRegex, (match, entityType) => {
    return `<div className="grid grid-cols-3 gap-2">
                                                <button
                                                    onClick={() => handleApproval((${entityType} as any).approvalId, 'APPROVE')}
                                                    disabled={isProcessing}
                                                    className="py-2 bg-[#29258D] text-white text-xs font-semibold rounded-md hover:bg-[#29258D]/90 transition-all flex items-center justify-center gap-1"
                                                >
                                                    {isProcessing ? <PiSpinner className="animate-spin" /> : <PiCheckCircle />} Approve
                                                </button>
                                                <button
                                                    onClick={() => handleApproval((${entityType} as any).approvalId, 'ADJUST')}
                                                    disabled={isProcessing}
                                                    className="py-2 bg-white border border-amber-200 text-amber-600 text-xs font-semibold rounded-md hover:bg-amber-50 transition-all flex items-center justify-center gap-1"
                                                >
                                                    {isProcessing ? <PiSpinner className="animate-spin" /> : <PiClock />} Adjust
                                                </button>
                                                <button
                                                    onClick={() => handleApproval((${entityType} as any).approvalId, 'REJECT')}
                                                    disabled={isProcessing}
                                                    className="py-2 bg-white border border-rose-200 text-rose-600 text-xs font-semibold rounded-md hover:bg-rose-50 transition-all flex items-center justify-center gap-1"
                                                >
                                                    {isProcessing ? <PiSpinner className="animate-spin" /> : <PiXCircle />} Reject
                                                </button>
                                            </div>`;
});


// We have list buttons of form: 
const listRegex = /<button onClick={\(\) => handleApproval\(\(([^ ]+) as any\)\.approvalId, 'REJECT'\)}\s*disabled={isProcessing}\s*className="(py-2\.5 px-4 bg-white border border-rose-200 text-rose-600[^"]+)"\s*>({isProcessing \? <PiSpinner className="animate-spin text-sm" \/> : <PiXCircle className="text-sm" \/>} Reject)<\/button>\s*<button onClick={\(\) => handleApproval\(\(\1 as any\)\.approvalId, 'APPROVE'\)}\s*disabled={isProcessing}\s*className="([^"]+)"\s*>({isProcessing \? <PiSpinner className="animate-spin text-sm" \/> : <PiCheckCircle className="text-sm" \/>} Approve)<\/button>/g;

content = content.replace(listRegex, (match, entityType, rejectClass, rejectInner, approveClass, approveInner) => {
    return `<button onClick={() => handleApproval((${entityType} as any).approvalId, 'REJECT')} disabled={isProcessing} className="${rejectClass}">${rejectInner}</button>
                                                                <button onClick={() => handleApproval((${entityType} as any).approvalId, 'ADJUST')} disabled={isProcessing} className="py-2.5 px-4 bg-white border border-amber-200 text-amber-600 text-[11px] font-semibold uppercase tracking-widest rounded-lg hover:bg-amber-50 transition-all flex items-center gap-1.5">{isProcessing ? <PiSpinner className="animate-spin text-sm" /> : <PiClock className="text-sm" />} Adjust</button>
                                                                <button onClick={() => handleApproval((${entityType} as any).approvalId, 'APPROVE')} disabled={isProcessing} className="${approveClass}">${approveInner}</button>`;
});

fs.writeFileSync(file, content);

File: scratch.py
import re

file_path = '/Users/mac/Desktop/pesanest1/src/components/dashboard/ApprovalQueue.tsx'

with open(file_path, 'r') as f:
    content = f.read()

# Replace state definition
content = content.replace(
    'const [isProcessing, setIsProcessing] = useState(false);',
    'const [processingAction, setProcessingAction] = useState<{id: string, action: string} | null>(null);'
)

# Replace handleApproval signature and body
content = content.replace(
    'setIsProcessing(true);',
    'setProcessingAction({id: approvalId, action});'
)

content = content.replace(
    'setIsProcessing(false);',
    'setProcessingAction(null);'
)

# Now, we need to replace button disabled and icon states.
# The `disabled={isProcessing}` should become `disabled={processingAction !== null}`
content = content.replace(
    'disabled={isProcessing}',
    'disabled={processingAction !== null}'
)

# And inside buttons, `{isProcessing ? <PiSpinner className="animate-spin" /> : <PiCheckCircle />} Approve`
# Need to replace {isProcessing ? ...} with {processingAction?.id === (entity as any).approvalId && processingAction?.action === 'ACTION' ? ...}
# Let's iterate using regex and replace `isProcessing ?` with the specific check.
# Since the buttons might be split across lines, let's craft targeted regexes.

# For Approve:
content = re.sub(
    r"onClick={\(\) => handleApproval\(\(([^ ]+) as any\)\.approvalId, 'APPROVE'\)}(.*?)disabled={processingAction !== null}(.*?)>(\s*)\{isProcessing \?",
    r"onClick={() => handleApproval((\1 as any).approvalId, 'APPROVE')}\2disabled={processingAction !== null}\3>\4{processingAction?.id === (\1 as any).approvalId && processingAction?.action === 'APPROVE' ?",
    content,
    flags=re.DOTALL
)

# For Reject:
content = re.sub(
    r"onClick={\(\) => handleApproval\(\(([^ ]+) as any\)\.approvalId, 'REJECT'\)}(.*?)disabled={processingAction !== null}(.*?)>(\s*)\{isProcessing \?",
    r"onClick={() => handleApproval((\1 as any).approvalId, 'REJECT')}\2disabled={processingAction !== null}\3>\4{processingAction?.id === (\1 as any).approvalId && processingAction?.action === 'REJECT' ?",
    content,
    flags=re.DOTALL
)

# For Adjust:
content = re.sub(
    r"onClick={\(\) => handleApproval\(\(([^ ]+) as any\)\.approvalId, 'ADJUST'\)}(.*?)disabled={processingAction !== null}(.*?)>(\s*)\{isProcessing \?",
    r"onClick={() => handleApproval((\1 as any).approvalId, 'ADJUST')}\2disabled={processingAction !== null}\3>\4{processingAction?.id === (\1 as any).approvalId && processingAction?.action === 'ADJUST' ?",
    content,
    flags=re.DOTALL
)


# There might be some list buttons where disabled and class are in a different order,
# wait, my regex: disabled={processingAction !== null} is in the middle. Let's make an easier regex.

content = re.sub(
    r"onClick={\(\) => handleApproval\(\(([^ ]+) as any\)\.approvalId, '([^']+)'\)}(.*?)\{isProcessing \?",
    r"onClick={() => handleApproval((\1 as any).approvalId, '\2')}\3{processingAction?.id === (\1 as any).approvalId && processingAction?.action === '\2' ?",
    content,
    flags=re.DOTALL
)

with open(file_path, 'w') as f:
    f.write(content)
